Userland/Libraries/LibJS/Runtime/TypedArrayPrototype.cpp

```diff
@@ -199,6 +268,215 @@ JS_DEFINE_NATIVE_FUNCTION(TypedArrayPrototype::for_each)
+    if (vm.exception())
+        return {};
+
+    auto value_n = Value(n);
+    if (value_n.is_positive_infinity())
+        return Value(false);
+    else if (value_n.is_negative_infinity())
+        n = 0;
+
+    u32 k;
+    if (n >= 0) {
+        k = n;
+    } else {
+        auto relative_k = length + n;
+        if (relative_k < 0)
+            relative_k = 0;
+        k = relative_k;
+    }
+
+    auto search_element = vm.argument(0);
+    for (; k < length; ++k) {
+        auto element_k = typed_array->get(k);
+
+        if (same_value_zero(search_element, element_k))
+            return Value(true);
+    }
+
+    return Value(false);
+}
+
+// 23.2.3.14 %TypedArray%.prototype.indexOf ( searchElement [ , fromIndex ] ), https://tc39.es/ecma262/#sec-%typedarray%.prototype.indexof
+JS_DEFINE_NATIVE_FUNCTION(TypedArrayPrototype::index_of)
+{
+    auto typed_array = typed_array_from(vm, global_object);
+    if (!typed_array)
+        return {};
+
+    auto length = typed_array->array_length();
+
+    if (length == 0)
+        return Value(-1);
+
+    auto n = vm.argument(1).to_integer_or_infinity(global_object);
+    if (vm.exception())
+        return {};
+
+    auto value_n = Value(n);
+    if (value_n.is_positive_infinity())
+        return Value(-1);
+    else if (value_n.is_negative_infinity())
+        n = 0;
+
+    u32 k;
+    if (n >= 0) {
+        k = n;
+    } else {
+        auto relative_k = length + n;
+        if (relative_k < 0)
+            relative_k = 0;
+        k = relative_k;
+    }
+
+    auto search_element = vm.argument(0);
+    for (; k < length; ++k) {
+        auto k_present = typed_array->has_property(k);
+        if (k_present) {
+            auto element_k = typed_array->get(k);
+
+            if (strict_eq(search_element, element_k))
+                return Value(k);
+        }
+    }
+
+    return Value(-1);
+}
+
+// 23.2.3.17 %TypedArray%.prototype.lastIndexOf ( searchElement [ , fromIndex ] ), https://tc39.es/ecma262/#sec-%typedarray%.prototype.lastindexof
+JS_DEFINE_NATIVE_FUNCTION(TypedArrayPrototype::last_index_of)
+{
+    auto typed_array = typed_array_from(vm, global_object);
+    if (!typed_array)
+        return {};
+
+    auto length = typed_array->array_length();
+
+    if (length == 0)
+        return Value(-1);
+
+    double n;
+    if (vm.argument_count() > 1) {
+        n = vm.argument(1).to_integer_or_infinity(global_object);
+        if (vm.exception())
+            return {};
+    } else {
+        n = length - 1;
+    }
+
+    if (Value(n).is_negative_infinity())
+        return Value(-1);
+
+    i64 k;
+    if (n >= 0)
+        k = min(n, (i64)length - 1);
+    else
+        k = length + n;
+
+    auto search_element = vm.argument(0);
+    for (; k >= 0; --k) {
+        auto k_present = typed_array->has_property(k);
+        if (k_present) {
+            auto element_k = typed_array->get(k);
+
+            if (strict_eq(search_element, element_k))
+                return Value((u32)k);
```

Comment:
Casting to `double` here seems safer, no?

Reply:
k is guaranteed to be non-negative here, so a cast to u32 is free here (unlike a double conversion)

Reply:
but is it also guaranteed to fit in an u32?

Reply:
actually, its guaranteed to fit in a i32 (Due to the 2GB TypedArray length limit). which means i can use a i32 here instead of a i64, will fix.

Reply:
done :)